Resfix respredict

sunal1996 opened the pull request:
We want to enable customized predictions coming out of TIMED. With this PR, the following can be rendered possible:

-Fixing certain aminoacid positions to the consensus aminoacid
-Predicting only specific aminoacid positions and reverting rest to the consensus sequence.
-Fix certain aminoacid positions to the desired aminoacid

Relevant README updates have been added.

universvm commented on design_utils/utils.py:
```diff
@@ -16,7 +19,239 @@
+)
+
+
+def constrain_aa_resnum(res_to_fix):
```
res_to_fix has no typing.

universvm commented on design_utils/utils.py:
```diff
@@ -16,7 +19,239 @@
+    return res_indexes
+
+
+def customize_fixed_residues(
+    pdb_to_sequence, pdb_to_real_sequence, chains_to_fix, res_to_fix, flat_dataset_map
+):
```
no typing

universvm commented on design_utils/utils.py:
```diff
@@ -16,7 +19,239 @@
+                            )
+
+
+def customize_predicted_residues(
+    pdb_to_sequence,
+    pdb_to_real_sequence,
+    chains_to_predict,
+    res_to_predict,
+    flat_dataset_map,
+):
```
no typing

universvm commented on predict.py:
```diff
@@ -21,13 +23,33 @@
 )
 
 
+def parse_residue_list(residue_list):
```
no typing

universvm commented on predict.py:
```diff
@@ -294,5 +334,30 @@ def main(args):
+        help="Specify the chains need to be predicted, separated by commas. Example: 'A, B' ",
+    )
+
+    parser.add_argument(
+        "--res_to_predict",
+        type=parse_residue_list,
+        help="Specify the residue numbers that need to be predicted, separated by commas. Example: '2 3 4 5, 6 7 8' will predict 2 3 4 5 6 7 8th residues and leave the rest as same as WT",
+    )
+
```
How is this different from res_to_fix?

In theory we should predict anything that is not res to fix. right?

universvm commented on predict.py:
```diff
@@ -294,5 +334,30 @@ def main(args):
         action="store_true",
         help="Whether the structure is NMR. NMR will have different states so TIMED will try to build a consensus",
     )
+
+    parser.add_argument(
+        "--res_to_fix",
+        type=parse_residue_list,
+        help="Specify the residue numbers that need to be fixed, separated by commas. Example: '2 3 4 5, 6 7 8' will keep 2 3 4 5 6 7 8th residues as identical with WT and predict the rest.",
+    )
+
+    parser.add_argument(
+        "--chains_to_fix",
+        type=parse_chain_list,
+        help="Specify the chains need to be fixed, separated by commas. Example: 'A, B' ",
+    )
+
+    parser.add_argument(
+        "--chains_to_predict",
+        type=parse_chain_list,
+        help="Specify the chains need to be predicted, separated by commas. Example: 'A, B' ",
+    )
+
```
Imo this is too many arguments which inherently make the code harder to understand. 

Let's stick with the "to_fix' and for the moment predict anything else.

Also I think it should be only one input argument rather than two. 

like


```
--res_to_fix A: 5, 10, 22; B: 2, 23, 1
```

universvm commented on design_utils/utils.py:
```diff
@@ -16,7 +19,239 @@
+
+def constrain_aa_resnum(res_to_fix):
+    """
+    At the moment the user inputs for residue fixing and residue prediction are taken as a string.
+    This function takes a user input and separates the integer bits from string bits. Particularly useful if the user wants to specify the aminoacid constraints when fixing residues.
+    For example, if the user puts --res_to_fix "10Y", this function will come in handy to constrain the 10th residue to a tyrosine.
+    If the user enters --res_to_fix "5", then the aminoacid will be considered "X" at the beginning, but this will not be inserted into the predictions function to create dictionaries for individual chains of proteins with their residue numbers.
```
This docstring does not explain what the function does. 

"At the moment" may change in the future and in fact, if anything changes it will confuse the user. Have a look at the other docstrings. Your docstring should describe what your function does with the input and what it returns.

universvm commented on design_utils/utils.py:
```diff
@@ -16,7 +19,239 @@
+
+     Parameters
+    ----------
+    res_to_fix: tuple
```
Tuple of what?

universvm commented on design_utils/utils.py:
```diff
@@ -16,7 +19,239 @@
+        constrained_resnum = []
+        constrained_aa = []
+        for i in item:
+            match = re.search(r"\d+", i)
```
What is this doing? Regex is hard to decypher without comments

universvm replied:
(add comments)

sunal1996 commented:
Fixed the types and added the exception handling to handle the following cases:

-If the user puts a chain that is not in the dataset
-If the user tries using --chains_to_fix w/o --res_to_fix or vice versa
-If the user tries using --chains_to_predict w/o res_to_predict or vice versa
-If the user tries predicting and fixing a residue number at the same time

At the moment, sadly, all of these exceptions seem to be handled after TIMED makes the predictions. Would be good to make it before the prediction starts, this would be a problem for big datasets with lots of structures in it.

universvm commented:
Closed in favor of other PR to come soon
